cpp/src/arrow/record_batch.cc

```diff
@@ -487,6 +487,27 @@ struct EnumeratedStatistics {
 };
 using OnStatistics =
     std::function<Status(const EnumeratedStatistics& enumerated_statistics)>;
+using ArrayStatisticsAndTypeVector =
+    std::vector<std::pair<std::shared_ptr<ArrayStatistics>, std::shared_ptr<DataType>>>;
+ArrayStatisticsAndTypeVector ExtractArrayDataAndType(const RecordBatch& record_batch) {
+  ArrayDataVector traverse;
+  ArrayStatisticsAndTypeVector result;
+  for (int i = 0; i < record_batch.num_columns(); ++i) {
+    auto column_array_data = record_batch.column(i)->data();
+    result.emplace_back(column_array_data->statistics, column_array_data->type);
```

Comment:
How about just returning `std::shared_ptr<ArrayData>` instead of `std::pair<std::shared_ptr<ArrayStatistics>, std::shared_ptr<DataType>>`?